src/expr/src/relation/func.rs

```diff
@@ -709,6 +709,103 @@ where
     })
 }
 
+// The expected input is in the format of [((OriginalRow, InputValue), OrderByExprs...)]
+fn first_value<'a, I>(
+    datums: I,
+    temp_storage: &'a RowArena,
+    order_by: &[ColumnOrder],
+    window_frame: &WindowFrame,
+) -> Datum<'a>
+where
+    I: IntoIterator<Item = Datum<'a>>,
+{
+    // Sort the datums according to the ORDER BY expressions and return the (OriginalRow, InputValue) record
+    let datums = order_aggregate_datums(datums, order_by);
+
+    // Decode the input (OriginalRow, InputValue) into separate datums
+    let datums = datums
+        .into_iter()
+        .map(|d| {
+            let mut iter = d.unwrap_list().iter();
+            let original_row = iter.next().unwrap();
+            let input_value = iter.next().unwrap();
+
+            (input_value, original_row)
+        })
+        .collect_vec();
+
+    let length = datums.len();
+    let mut result: Vec<(Datum, Datum)> = Vec::with_capacity(length);
+    for (idx, (current_datum, original_row)) in datums.iter().enumerate() {
+        let first_value = match &window_frame.start_bound {
+            // Always return the current value
+            WindowFrameBound::CurrentRow => *current_datum,
+            WindowFrameBound::UnboundedPreceding => {
+                if let WindowFrameBound::OffsetPreceding(end_offset) = &window_frame.end_bound {
+                    let end_offset = usize::try_from(*end_offset)
+                        .expect("Window frame offset does not fit in usize");
+
+                    // If the frame ends before the first row, return null
+                    if idx < end_offset {
+                        Datum::Null
+                    } else {
+                        datums[0].0
+                    }
+                } else {
+                    datums[0].0
+                }
+            }
+            WindowFrameBound::OffsetPreceding(offset) => {
+                let start_offset =
+                    usize::try_from(*offset).expect("Window frame offset does not fit in usize");
+                let start_idx = idx.saturating_sub(start_offset);
+                if let WindowFrameBound::OffsetPreceding(end_offset) = &window_frame.end_bound {
+                    let end_offset = usize::try_from(*end_offset)
+                        .expect("Window frame offset does not fit in usize");
+
+                    // If the frame is empty or ends before the first row, return null
+                    if start_offset < end_offset || idx < end_offset {
+                        Datum::Null
+                    } else {
+                        datums[start_idx].0
+                    }
+                } else {
+                    datums[start_idx].0
+                }
+            }
+            WindowFrameBound::OffsetFollowing(offset) => {
+                let start_offset =
+                    usize::try_from(*offset).expect("Window frame offset does not fit in usize");
+                let start_idx = idx + start_offset;
```

Comment:
worth doing some kind of checked_add here (and...uh...dunno how you error there) in case a user specifies an offset that is u64::max. Probably need to add some tests for these too.

Reply:
Ah, good catch. I think `saturating_add` would be good enough in this case, as any offsets this high will almost always be out of bounds, since I think something else would overflow before you tried to have a collection that large (`u64::MAX` bytes is about 18 exabytes).

Also going to add some tests to exercise this behavior!

BTW, since we're using u64s, we support far higher numbers than Postgres, which uses an `i64` for the offset, and just fails to parse it if it's too large. Should we also reject these offsets (between `i64::MAX` and `u64::MAX`) outright or keep the current behavior?

Reply:
Pushed a new version with `saturating_add` + added tests for overflow.